lib/modules.nix

```diff
@@ -807,6 +807,9 @@ let
 
     in warnDeprecation opt //
       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
+        # The merged value before applying the options `apply` function to it.
+        # In general though, `apply` should not be used, it's an anti-pattern.
+        beforeApply = res.mergedValue;
```

Comment:
> What if `opt.beforeApply` is wrong

Not sure what you mean by that?

> and a non-trivial number of modules relies on opt.beforeApply

It only makes sense to rely on `beforeApply` if the option has an `apply`, otherwise it's the same as the options `value`. And `apply` is being discouraged already, you won't find much new uses. The main use case here really is just allowing certain reasonable configs for older options that already use `apply` and can't be migrated away from easily.

I agree that it's a high cost to pay for very little value though..

Reply:
This is also a good time to mention #299736 as an alternative implementation that doesn't add a new attribute to all options.

Reply:
> > What if `opt.beforeApply` is wrong
> 
> Not sure what you mean by that?

This time `opt.value` is wrong. Next time `opt.beforeApply` is wrong. Why would this sequence stop?



> And `apply` is being discouraged already, you won't find much new uses.

I'm not convinced of that.

Reply:
`opt.value` is "wrong" this time because `apply` exists, so we need `beforeApply`. The sequence stops here because we don't have a "pre-apply" step. After the values get merged, it gets passed to `apply` directly.

> I'm not convinced of that.

At least I'm heavily discouraging `apply` whenever I see it :)

Reply:
Hmm, I might have overreacted.

Bloat argument still stands though, or are we betting on _replacing_ `evalModules` in the foreseeable future?
I like the idea of having time to do that... :)